src/OrchardCore.Modules/OrchardCore.Layers/Services/LayerFilter.cs

```diff
@@ -124,18 +125,21 @@ public async Task OnResultExecutionAsync(ResultExecutingContext context, ResultE
                 if (widgetDefinitions.TryGetValue(widget.ContentItem.ContentType, out var definition) &&
                     (!definition.IsSecurable() || await _authorizationService.AuthorizeAsync(context.HttpContext.User, CommonPermissions.ViewContent, widget.ContentItem)))
                 {
-                    var widgetContent = await _contentItemDisplayManager.BuildDisplayAsync(widget.ContentItem, updater);
+                    // Note: We clone the cached content item to avoid sharing the same instance across threads when rendering widgets.
```

Comment:
Everything with the fix looks good. However, I thought I'd ask what you think about line 81 which looks roughly like this

```cs
if (!_memoryCache.TryGetValue(...))
{
    _memoryCache.Set(...)
}
```

Seems like this is also prone to race conditions. Two threads may check the cache, find that the value is missing, and try to add it. However, I think the side effects of this is much less severe. At worst, the second thread will duplicate the work of the first, but I don't think that causes any errors down the line. Just a bit of wasted work.

It not be worth changing anything here, just something I noticed.

Reply:
Yes, that is fine and often done this way. It doesn't matter if the cache entry gets replaced by another one, in the rare case when this is hit more than once.